packages/worker/src/worker.ts

```diff
@@ -569,11 +564,8 @@ export class Worker {
       if (isCodeBundleOption(compiledOptions.workflowBundle)) {
         return compiledOptions.workflowBundle;
       } else if (isPathBundleOption(compiledOptions.workflowBundle)) {
-        const [code, sourceMap] = await Promise.all([
-          fs.readFile(compiledOptions.workflowBundle.codePath, 'utf8'),
-          fs.readFile(compiledOptions.workflowBundle.sourceMapPath, 'utf8'),
-        ]);
-        return { code, sourceMap };
+        const code = await fs.readFile(compiledOptions.workflowBundle.codePath, 'utf8');
```

Comment:
Does it matter if we use the path here as name of the vm script?

Reply:
At this point, I don't think supporting this use case is worth delaying a release.
But should we want to do it, this would probably require rewriting the content of the `sourceMapURL` so that the filename matches the actual file.

Reply:
Sounds like it shouldn’t matter

Reply:
> I would like to see if there’s a way to get rid of the global script and context, maybe just store on the worker for now?
> Wondering why it’s needed and if there’s a way around it.

The issue is that parsing of the source map is done by the debugger (ie. in VS Code's process), concurrently with the execution of the workflow code (ie. node's process). Parsing the source map takes some time, and is associated with the vm context in which the script was discovered. Once parsing is completed, the debugger informs node about file mappings that it has discovered, _with the associated script id and context id_. Now, if that context has already been garbaged collected at that point (which is very likely with workflows since activations are very short lived and workflow bundles tends to be rather large), then node throws an error, ignoring file mappings provided by the debugger.

Adding preload and caching of the script and context has decreased breakpoints misses from over 9 out of 10 in some test scenarios to no miss at all. Workflow breakpoints are simply not reliable without this.

We can indeed attach the objects to the worker instead of keeping it in a global variable, and that was actually my original intent. Another possibility I'm considering at the moment is to attach the objects to a 10 seconds timer. That is more than enough to let the debugger completes parsing. I will have to look at vscode's debugger trace log however to determine if this cause errors at some later point.

I also observed that preloading is more reliable if performed _before_ worker threads are started, rather that once in worker threads. For example, I had cases where only one or two threads (over 8) would hit breakpoints, until disable and reenable all breakpoints. Looking at a trace from vscode's debugger, it looks like the debugger attaches some listeners to newly created threads, so my hypothesis is that the debugger might miss some type of events if they happens very early after thread creation.

Reply:
My intuition also tells me that it should not matter (after all, the code being executed gave that source map, so why would we care about the name of the file), but experimentation tells me that setting breakpoints from the UI doesn't work unless the file name stored inside the source map exactly matches the script's filename.

TBH, that's not a big thing to fix. I'm just concerned that it might delay a release. Your call.

Reply:
Done.